[WIP] Further development on intel_mp driver

takikawa opened the pull request:
The initial work on the `intel_mp` driver is now pending for some upstream review. This PR is for tracking additional work on the driver (keeping it on a separate branch so that the initial PR doesn't just keep growing) and making sure that CI passes.

Progress on this branch:
 - [x] pool mirroring for VMDq
 - [x] basic performance improvements for `intel_mp`
 - [x] improvements for debugging
 - [x] support for setting counters for queues (Tx and Rx)
 - [x] ported priority/rate limiting code from intel10g
 - [x] fixed some buggy app teardown code
 - [ ] more performance improvements
 - [x] make rx/tx queue numbers relative to the selected pool number (on other branch)

lukego committed:
net_device: Eliminate fast-path table allocations

Tables of callback functions are now cached as instance variables
rather than being created with {...} on every call. These allocations
were not being sunk and so this eliminates heap allocation & GC.

(cherry picked from commit 76782f505664845756e2a58b4c4751d6312d087b)

vmaffione committed:
apps: vhost: README.md: fix typo

wingo committed:
Skip lib.multi_copy selftest if AVX2 unavailable

wingo committed:
Merge branch 'lwaftr' into lwaftr-2017-04-24

takikawa committed:
First attempt at VMDq for intel_mp

lukego committed:
Merge #1127 branch 'vmaffione/docfix' into next

lukego committed:
Merge #1126 branch 'lukego/dealloc' into next

takikawa committed:
Fix up VMDq mode so that it actually runs

takikawa committed:
First attempt at tests for VMDq mode

takikawa committed:
Try testing different MAC addrs with VMDq

takikawa committed:
Fix MAC address registration in VMDq mode

Avoid using index sets and just query the registers since it's
hard to share this state among the apps. Also do the MAC registration
in the right place.

takikawa committed:
Add comment explaining some RSS code

takikawa committed:
Enable RSS queues via PSRTYPE properly for VMDq

takikawa committed:
Use :bits to set the VMDq RSS mode properly

takikawa committed:
Remove RSS bit check for setting RETA

RSS is always enabled, so there shouldn't be a case in which
this is a no-op anyway. Also the check is complicated by the fact
that RSS can be enabled in many ways (with DCB, VMDq, etc.).

takikawa committed:
Fix RTTPCS bit set that had a wrong length arg

wingo committed:
Add soaktest to "snabb lwaftr" usage.

takikawa committed:
Adjust test to test that both vmdq pools get pkts

takikawa committed:
Move MAC pool enable code and make it cross-NIC

Also add code to unset MAC on stop()

takikawa committed:
Adjust the VMDq initialization assertions

takikawa committed:
Error when the max number of MAC addresses is reached

takikawa committed:
Add code for enabling VLAN filtering/tagging

This code is mostly copied from the intel10g.lua driver.
Also add a new VLAN test and adjust old tests as needed.

takikawa committed:
Refactor vmdq test script

The previous iteration was unreliable for unknown reasons. This
refactoring should be clearer and run more reliably.

takikawa committed:
Enable VLAN takedown code for VMDq

wingo committed:
Disable bus-mastering cleanup for intel_mp

See https://github.com/snabbco/snabb/issues/1145 for more discussion.

wingo committed:
Merge pull request #809 from Igalia/no-supervisor-cleanup-for-intel_mp

Disable bus-mastering cleanup for intel_mp

lukego committed:
README.md: Document core.worker multiprocessing

lukego committed:
core.worker: Remove dead code

lukego committed:
Merge #1021 branch 'lukego/multiprocess-engine' into integrate-multiproc-lwaftr

lukego committed:
Merge #1133 branch 'igalia/lwaftr-2017-04-24' into integrate-multiproc-lwaftr

takikawa committed:
Refactor intel_mp recv test scripts to reduce duplication

takikawa committed:
Enable VLAN tag stripping in VMDq mode

Also add new test for VLAN stripping

takikawa committed:
Adjust VLAN test to also test tag insertion

takikawa committed:
VMDq settings for transmit queues

takikawa committed:
Add a test for VLAN tag insertion & VMDq tx

takikawa committed:
Set registers needed to make Tx test work

 * PFVFTE to enable Tx from appropriate VFs
 * RTTD1TC for bandwidth allocation algorithm for Tx

takikawa committed:
Fix assertion for Rx queue num to allow nil

takikawa committed:
Adjust intel_mp to error with VMDq on non-82599

lukego committed:
README.md: Fix multiprocessing diagram syntax

lukego committed:
Merge #1148 branch 'lukego/integrate-multiproc-lwaftr' into next

takikawa committed:
Remove unnecessary helper function

takikawa committed:
Fix permissions on two intel_mp tests

takikawa committed:
Fix 1q vmdq test and testrecv.lua

The former wasn't updated for an API change and the latter
was missing a module declaration.

takikawa committed:
Adjust VLAN test to avoid testing tag insertion

The tag insertion part of this test didn't work because the
MAC didn't match when the app tried to resend the packet. It's
not necessary since the transmit test covers this anyway.

takikawa committed:
Add waits to avoid race conditions in vmdq tests

takikawa committed:
Add documentation for new vmdq-related parameters

takikawa committed:
Refactor vmdq-related assertions

Commit:
Make intel_mp tests run even when either 1G or 10G cards are not present

takikawa committed:
Adjust VMDq test for 82599 to be more robust

There are multiple ids for 82599 variants that may
be added later

takikawa committed:
Add VMDq consistency check for intel_mp apps

Makes sure that all apps on a given NIC either use or don't
use VMDq uniformly

Commit:
Disable VMDq check if not in VMDq mode

takikawa committed:
Allow mirroring in VMDq mode

Ported mirroring from intel10g driver. No VLAN mirroring implemented
yet. Also added a test for mirroring.

takikawa committed:
Avoid use of `bits` function in hot code

The `bits` function is slow because of the call to
`pairs` inside of it.

takikawa committed:
Avoid `pairs` in more places in intel_mp

teknico commented on src/apps/intel_mp/intel_mp.lua:
```diff
@@ -1047,14 +1047,13 @@ function Intel1g:init ()
 function Intel1g:link_status ()
-   local mask = bits { Link_up = 1 }
-   return bit.band(self.r.STATUS(), mask) == mask
+   return bit.band(self.r.STATUS(), lshift(1, 1)) == mask
```
On the right side you're still using `mask`, which is not defined anymore.

takikawa replied:
Whoops, nice catch. This is the problem with code that you can't test. :p

takikawa committed:
Fix mistake in commit d1b0a87fb23ce9b49a

takikawa committed:
Sort keys in debug prinouts for easier reading

takikawa committed:
Fix offsets for QBTC/QPTC registers

See page 705 in the 82599 datasheet

takikawa committed:
Add rxcounter support in intel_mp

Allows queue counters to be assigned to app instances
as the intel10g driver allows. Also implements get_rxstats().

For 1g NICs, get_rxstats() gets the stats for the
(non-user assigned) per-queue registers.

takikawa committed:
Add txcounter support in intel_mp

Similar to the rxcounter support. This also enables counter
sync in the `push` method so that statistics are available
when only Tx is enabled for the app.

takikawa committed:
Port Tx rate limit and priority code to intel_mp

teknico commented on src/apps/intel_mp/intel_mp.lua:
```diff
@@ -324,7 +326,9 @@ function Intel:new (conf)
       rxcounter = conf.rxcounter,
-      txcounter = conf.txcounter
+      txcounter = conf.txcounter,
+      limit = conf.rate_limit,
```
I'd rather keep calling this `rate_limit`, because just `limit` is unclear. What do you think?

takikawa replied:
I like your suggestion, I'll change it.

takikawa committed:
Fix silly typo in method call

takikawa committed:
Add code from intel10g for discarding unsent pkts

Replaces crashing code that was just using packet.free
on all of the txqueue elements.

takikawa committed:
Fix some buggy app teardown code in intel_mp

takikawa committed:
Fix teardown test for this branch

The test depended on some modifications from another branch
that disabled app link presence tests.

takikawa committed:
Rename `limit` field to `rate_limit`

takikawa committed:
Fix another typo in VLAN code

takikawa commented:
I'm going to close this since we are no longer doing development on this branch.
